processor/samplingprocessor/processor.go

```diff
@@ -18,79 +18,192 @@ import (
-					return sp.sampleFunc()
-				})
-			}
+	}
+	// Drop based on ratio and condition
+	for i := 0; i < ld.ResourceLogs().Len(); i++ {
+		for j := 0; j < ld.ResourceLogs().At(i).ScopeLogs().Len(); j++ {
+			ld.ResourceLogs().At(i).ScopeLogs().At(j).LogRecords().RemoveIf(func(logRecord plog.LogRecord) bool {
+				logCtx := ottllog.NewTransformContext(
+					logRecord,
+					ld.ResourceLogs().At(i).ScopeLogs().At(j).Scope(),
+					ld.ResourceLogs().At(i).Resource(),
+					ld.ResourceLogs().At(i).ScopeLogs().At(j),
+					ld.ResourceLogs().At(i),
+				)
+				match, err := sp.condition.Match(ctx, logCtx)
+				return err == nil && match && sampleFunc(sp.dropCutOffRatio)
+			})
 		}
-		return ld, nil
 	}
+	return ld, nil
 }
 
-func (sp *samplingProcessor) processMetrics(_ context.Context, md pmetric.Metrics) (pmetric.Metrics, error) {
-	switch {
-	case sp.dropCutOffRatio == 1.0: // Drop everything
-		return pmetric.NewMetrics(), nil
-	case sp.dropCutOffRatio == 0.0: // Drop nothing
+func (sp *metricsSamplingProcessor) processMetrics(ctx context.Context, md pmetric.Metrics) (pmetric.Metrics, error) {
+	// Drop nothing
+	if sp.dropCutOffRatio == 0.0 {
 		return md, nil
-	default: // Drop based on ratio
-		for i := 0; i < md.ResourceMetrics().Len(); i++ {
-			for j := 0; j < md.ResourceMetrics().At(i).ScopeMetrics().Len(); j++ {
-				md.ResourceMetrics().At(i).ScopeMetrics().At(j).Metrics().RemoveIf(func(_ pmetric.Metric) bool {
-					return sp.sampleFunc()
+	}
+	// Drop based on ratio and condition
+	for i := 0; i < md.ResourceMetrics().Len(); i++ {
+		for j := 0; j < md.ResourceMetrics().At(i).ScopeMetrics().Len(); j++ {
+			md.ResourceMetrics().At(i).ScopeMetrics().At(j).Metrics().RemoveIf(func(metric pmetric.Metric) bool {
+				anyMatch := false
+				eachDatapoint(metric, func(dp any) {
+					datapointCtx := ottldatapoint.NewTransformContext(
+						dp,
+						metric,
+						md.ResourceMetrics().At(i).ScopeMetrics().At(j).Metrics(),
+						md.ResourceMetrics().At(i).ScopeMetrics().At(j).Scope(),
+						md.ResourceMetrics().At(i).Resource(),
+						md.ResourceMetrics().At(i).ScopeMetrics().At(j),
+						md.ResourceMetrics().At(i),
+					)
+					curDatapointMatched, err := sp.condition.Match(ctx, datapointCtx)
+					anyMatch = anyMatch || (err == nil && curDatapointMatched)
 				})
-			}
+
+				// if no datapoints matched, check if metric matches without a datapoint
+				if !anyMatch {
+					metricCtx := ottldatapoint.NewTransformContext(
+						nil,
+						metric,
+						md.ResourceMetrics().At(i).ScopeMetrics().At(j).Metrics(),
+						md.ResourceMetrics().At(i).ScopeMetrics().At(j).Scope(),
+						md.ResourceMetrics().At(i).Resource(),
+						md.ResourceMetrics().At(i).ScopeMetrics().At(j),
+						md.ResourceMetrics().At(i),
+					)
+					metricMatch, err := sp.condition.Match(ctx, metricCtx)
+					anyMatch = err == nil && metricMatch
+				}
+
+				return anyMatch && sampleFunc(sp.dropCutOffRatio)
+			})
 		}
-		return md, nil
+	}
+	return md, nil
```

Comment:
Because we filter whole metrics, not datapoints, we should use the metric context and not the datapoint context here. That probably requires extending the `github.com/observiq/bindplane-agent/expr` package, but it'll make more sense to do it that way.

Reply:
Unfortunate timing on the merge 😳 great catch, I just opened a new PR to address your feedback: #1878